src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -181,127 +78,79 @@ let rec trace_split (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < lengt
-  : Lemma
-    (ensures (
-      let (tr1, _, _) = trace_split tr ts in
-      length tr1 = ts /\
-      get_entry_at tr1 i == get_entry_at tr i
-    ))
-  = let Snoc hd entry = tr in
-    if ts = last_ts tr then ()
-    else trace_split_left hd ts i
-
-let rec trace_split_mid (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < length tr})
-  : Lemma
-    (ensures (
-      let (_, e, _) = trace_split tr ts in
-      e == get_entry_at tr ts
-    ))
-  = let Snoc hd entry = tr in
-    if ts = last_ts tr then ()
-    else trace_split_mid hd ts
-
-let rec trace_split_right (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < length tr}) (i:timestamp{ts < i /\ i < length tr})
+let rec trace_split_get_entry (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < length tr}) (i:timestamp{i < length tr})
   : Lemma
-    (ensures (
-      let (_, _, tr2) = trace_split tr ts in
-      length tr2 = (length tr) - ts - 1 /\
-      get_entry_at tr2 (i - ts - 1) == get_entry_at tr i
-    ))
-  = let Snoc hd entry = tr in
-    if ts = last_ts tr then ()
-    else begin
-      trace_split_length tr ts;
-      if i < length hd then
-        trace_split_right hd ts i
-      else ()
-    end
+    (
+      let (tr1, e, tr2) = trace_split tr ts in
+      get_entry_at tr i == (
+        if i < ts then get_entry_at tr1 i
+        else if i = ts then e
+        else get_entry_at tr2 (i - ts - 1)
+      )
+    )
+  = norm_spec [zeta; delta_only [`%trace_split]] (trace_split #label_t);
+    let Snoc hd entry = tr in
+    if ts = last_ts tr || i >= length hd then ()
+    else trace_split_get_entry hd ts i
 
 let rec trace_split_matches_prefix (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < length tr})
   : Lemma
-    (ensures (
+    (
       let (tr1, e, _) = trace_split tr ts in
       tr1 == prefix tr ts /\
       (Snoc tr1 e) == prefix tr (ts + 1)
-    ))
-  = let Snoc hd entry = tr in
-    normalize_term_spec (prefix #label_t);
+    )
+  = norm_spec [zeta; delta_only [`%trace_split]] (trace_split #label_t);
+    norm_spec [zeta; delta_only [`%prefix]] (prefix #label_t);
+    let Snoc hd entry = tr in
     if ts = last_ts tr then ()
     else trace_split_matches_prefix hd ts
 
-let trace_split_grows (#label_t:Type) (tr:trace_ label_t) (ts:timestamp {ts < length tr})
-  : Lemma
-    (ensures (
-      let (tr1, e, _) = trace_split tr ts in
-      tr1 <$ tr /\
-      (Snoc tr1 e) <$ tr
-    ))
-  = trace_split_matches_prefix tr ts
-
 let rec trace_split_fmap (#a #b:Type) (tr:trace_ a) (ts:timestamp {ts < length tr}) (f:a -> b)
   : Lemma
-    (ensures (
+    (
       let (tr1, e, tr2) = trace_split tr ts in
       let (tr1', e', tr2') = trace_split (fmap_trace f tr) ts in
       fmap_trace f tr1 == tr1' /\
       fmap_trace_entry f e == e' /\
       fmap_trace f tr2 == tr2'
-    ))
-  = let Snoc hd entry = tr in
+    )
+  = norm_spec [zeta; delta_only [`%trace_split]] (trace_split #a);
+    norm_spec [zeta; delta_only [`%trace_split]] (trace_split #b);
```

Comment:
With functions that take implicit parameters, I find myself needing to do this kind of thing often, normalizing multiple times, once per applicable type. Is there a way to avoid this?

Reply:
I don't know about a better way, I agree this is unfortunate.